### PR TITLE
Khasa22 patch 1

### PR DESCRIPTION
**Pull request by khasa22**
network delay equation: double summation(line 35,65) ...
Have confusion how exactly encode it

### CODE REVIEW

**Comment by khasa22** [comment]
#

File: res-opt.mod [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#set of VNFs
+set VNF{S}; 
+#total computational resources
+param comput_res{n in N};	
+#VNF computational resources
+param c { v in S};  
+#throughtput between VFs links (v1,v2) 
+param vf_throughput{ES};
+#total bandwidth of (n1,n2)	
+param Pdelay{s1 in S,vf in VNF[s1]}:= sum {(v1,v2)in ES} vf_throughput[v1,v2]*c[vf];
```
Comment (MartinPJorge) [inline res-opt.mod]: What is this parameter?

You are summing over the multiplication of `vf_throughput[v1,v2]*c[vf]` which has units `[Mbps]*CPU`.
Such product cannot give a delay and we don't use this anywhere in the formulation, right?

File: res-opt.mod [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#network delay of individual links at E(G)
+param Ndelay{(n1,n2) in E}:= sum {(v1,v2) in ES}d[n1,n2]+qd[n1,n2];
+#total service delay
+param DS{S};
+param signal_strength{i in r, p in R};
+param background_noise;
+# transmission gain
+param T{i in r, p in R}:=throughput[i,p]*(log (1+(signal_strength[i,p]/background_noise)));
+param attachment{r,R} binary;
```
Comment (MartinPJorge) [inline res-opt.mod]: The radio attachment should be a decision variable, not a parameter.

File: res-opt.mod [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#param P {s1 in S,n in N} binary;
+
+# set of VFs hosted at node n
+var avf {s1 in S,vf in VNF[s1],n in N} binary;
+#set of VL hosted at node(n1,n2)
+var avl{ES,E} binary ;
+#var indicator_func {a[n] ,r[i]};
+maximize resources: sum {n in N} comput_res[n]-
+sum {s1 in S,n in N}
+        (sum {vf in VNF[s1]} if avf[s1,vf,n]==1 then c[vf])
```
Comment (MartinPJorge) [inline res-opt.mod]: You can replace the if statement as follows:
```ampl
sum {vf in VNF[s1]} avf[s1,vf,n] * c[s1, vf]
```

File: res-opt.mod [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+set w ;
+#servers
+set s;
+#hardware graph V(G) nodes (i.e. PoA,robot,switches,servers)
+set N := {R union r union w union s};
+# graph edges of hardware graph
+set E:={n1 in N, n2 in N}; 
+# service graph
+set S; 
+# vertices (VNFs) of service graph
```
Comment (MartinPJorge) [inline res-opt.mod]: These are the edges of the service graph.

I'd rename it as `VLs`.

Comment (MartinPJorge) [reply res-opt.mod]: On top, the edges should be defined for a particular service graph. Refactor `set ES ...` to be:
```ampl
set VLs{s in S} := {v1 in VNF[s], v2 in VNF[s]};
```

Comment (MartinPJorge) [reply res-opt.mod]: Update accordingly lines below to be consistent with the new definition.

File: res-opt.mod (hunk 1 above)
Comment (MartinPJorge) [reply res-opt.mod]: Although we do, we should change it to the avg. latency of an M/M/1 system -- see the overleaf.

File: res-opt.mod [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# transmission gain
+param T{i in r, p in R}:=throughput[i,p]*(log (1+(signal_strength[i,p]/background_noise)));
+param attachment{r,R} binary;
+#placement variable
+#param P {s1 in S,n in N} binary;
+
+# set of VFs hosted at node n
+var avf {s1 in S,vf in VNF[s1],n in N} binary;
+#set of VL hosted at node(n1,n2)
+var avl{ES,E} binary ;
```
Comment (MartinPJorge) [inline res-opt.mod]: You have to redefine it to consider the service. Something like:
```ampl
var avl {Si in S, VLs[Si], E} binary;
```

File: res-opt.mod [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#var indicator_func {a[n] ,r[i]};
+maximize resources: sum {n in N} comput_res[n]-
+sum {s1 in S,n in N}
+        (sum {vf in VNF[s1]} if avf[s1,vf,n]==1 then c[vf])
++ sum {(n1, n2) in E} throughput[n1,n2]-
+sum {(v1,v2) in ES}
+        (sum {s1 in S,vf in VNF[s1],n in N} if avf[s1,vf,n]==1 then vf_throughput[v1,v2]);
```
Comment (MartinPJorge) [inline res-opt.mod]: Replace this summation over virtual links by:

Same as above, replace the if statement using the product:
```ampl
sum {Si in S} sum {(v1,v2) in VLs[Si]}  avl[Si, v1,v2, n1,n2] * vf_throughput[Si, v1,v2];
```

File: res-opt.mod [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+subject to Consrtaint{i in r}:
+                                        sum {p in R} attachment[i,p]=1;
+subject to palcement{s1 in S, n in N}:
+                                    sum {v in VNF[s1]} avf[s1,v,n]*c[v] <= comput_res[n];
```
Comment (MartinPJorge) [inline res-opt.mod]: The placement constraint should apply to all services deployed within the node, i.e.:
```ampl
subject to placement {n in N}:
  sum {Si in S} sum {v in VNF[Si]} avf[Si,v,n] * c[Si,v] <= comput_res[n];
```

File: res-opt.mod [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+# service graph
+set S; 
+# vertices (VNFs) of service graph
+set ES:={v1 in S,v2 in S}; 
+#set of VNFs
+set VNF{S}; 
+#total computational resources
+param comput_res{n in N};	
+#VNF computational resources
+param c { v in S};
```
Comment (MartinPJorge) [inline res-opt.mod]: A VNF is not within the set of services, but within the set of VNFs of a particular service, therefore, it should be something like:
```ampl
param c {Si in S, v in VNFs[Si] };
```

File: res-opt.mod [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+# vertices (VNFs) of service graph
+set ES:={v1 in S,v2 in S}; 
+#set of VNFs
+set VNF{S}; 
+#total computational resources
+param comput_res{n in N};	
+#VNF computational resources
+param c { v in S};  
+#throughtput between VFs links (v1,v2) 
+param vf_throughput{ES};
```
Comment (MartinPJorge) [inline res-opt.mod]: Same for the throughput, to obtain a set of VLs from a service we need which service `Si` we talk about:
```ampl
param vf_throughput { Si in S, VLs[Si] };
```

File: res-opt.mod [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        (sum {vf in VNF[s1]} if avf[s1,vf,n]==1 then c[vf])
++ sum {(n1, n2) in E} throughput[n1,n2]-
+sum {(v1,v2) in ES}
+        (sum {s1 in S,vf in VNF[s1],n in N} if avf[s1,vf,n]==1 then vf_throughput[v1,v2]);
+subject to Consrtaint{i in r}:
+                                        sum {p in R} attachment[i,p]=1;
+subject to palcement{s1 in S, n in N}:
+                                    sum {v in VNF[s1]} avf[s1,v,n]*c[v] <= comput_res[n];
+subject to capacity{(n1,n2) in E}:
+                                        sum {(v1, v2) in ES} vf_throughput[v1,v2] <=throughput[n1,n2];
```
Comment (MartinPJorge) [inline res-opt.mod]: Here we miss to iterate over all possible services, as in the placement constraint above:
```ampl
                             sum {Si in S} sum {(v1,v2) in VLs[Si]}  vf_throughput[Si, v1,v2] <=throughput[n1,n2];
```

File: res-opt.mod [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+subject to PoA_feasiblity{i in r, p in R}:
+                                        sum {(v1,v2) in ES} vf_throughput[v1,v2]<= T[i,p];
```
Comment (MartinPJorge) [inline res-opt.mod]: As above, we miss the service indexing:
```ampl
subject to PoA_feasiblity{i in r, p in R}:
                                        sum {Si in S} sum {(v1,v2) in VL[Si]} vf_throughput[Si, v1,v2]<= T[i,p];
```

File: res-opt.mod [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+subject to delay{j in S}:
+                        sum {v in VNF[j]} Pdelay[j,v] + sum {(v1,v2) in ES} Ndelay[v1,v2] <=DS[j];
```
Comment (MartinPJorge) [inline res-opt.mod]: Use the service indexing to iterate over the VLs:
```ampl

subject to delay{j in S}:
                        sum {v in VNF[j]} Pdelay[j,v] + sum {(v1,v2) in VLs[j]} Ndelay[v1,v2] <=DS[j];
```

**Review by MartinPJorge — CHANGES_REQUESTED** [review]
On top of the comments, I would suggest that we use same notation as in the paper.

For example, rather than `throughput` we should use `lambda`. That'd make easier to follow the problem formulation and AMPL encoding. You can use comments to clarify what lambda is.

**Pull request merged**